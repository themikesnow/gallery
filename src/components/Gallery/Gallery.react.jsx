import React, { Component } from 'react';

import SearchInput from '../SearchInput/SearchInput.react';
import Image from './Image.react';
import Slider from './Slider.react';
import ImageProps from '../../constants/PropTypes';

export default class Gallery extends Component {
  static propTypes = {
    images: React.PropTypes.arrayOf(React.PropTypes.shape(ImageProps)),
    selectedImage: React.PropTypes.number,
    isPreviousEnabled: React.PropTypes.bool,
    isNextEnabled: React.PropTypes.bool,
    isBusy: React.PropTypes.bool,
    onSearch: React.PropTypes.func.isRequired,
    onPreviousImage: React.PropTypes.func.isRequired,
    onNextImage: React.PropTypes.func.isRequired,
    onSelectImage: React.PropTypes.func.isRequired,
  };

  static defaultProps = {
    images: null,
    selectedImage: 0,
    isPreviousEnabled: false,
    isNextEnabled: false,
  };

  static defaultProps = {
    isBusy: false,
  };

  render() {
    return (
      <div className="gallery">
        <div className="col-centered REVIEW">
          <SearchInput
            onSearch={this.props.onSearch}
            isBusy={this.props.isBusy}
          />
        </div>

        {this.props.images &&
          <div>
            {this.props.images.length > 0 ?
              <div>
                <Slider
                  image={this.props.images[this.props.selectedImage]}
                  isPreviousEnabled={this.props.isPreviousEnabled}
                  isNextEnabled={this.props.isNextEnabled}
                  onPreviousImage={this.props.onPreviousImage}
                  onNextImage={this.props.onNextImage}
                />
                <div className="images">
                  {this.props.images.map((i, index) => (
                    <Image
                      id={`image-${index}`}
                      key={index} // eslint-disable-line react/no-array-index-key
                      image={i}
                      index={index}
                      isActive={this.props.selectedImage === index}
                      onSelectImage={this.props.onSelectImage}
                    />
                  ))}
                </div>
              </div>
            : <div className="alert">
                No Images found...
              </div>
            }
          </div>
        }
      </div>
    );
  }
}
