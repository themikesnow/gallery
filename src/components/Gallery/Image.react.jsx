import React from 'react';
import classnames from 'classnames';
import ImageProps from '../../constants/PropTypes';

export default class Image extends React.Component {

  static propTypes = {
    id: React.PropTypes.string.isRequired,
    image: React.PropTypes.shape(ImageProps).isRequired,
    index: React.PropTypes.number,
    isActive: React.PropTypes.bool.isRequired,
    onSelectImage: React.PropTypes.func.isRequired,
  };

  static defaultProps = {
    index: 0,
  };

  constructor(props) {
    super(props);
    this.onSelectImage = ::this.onSelectImage;
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.isActive && (this.image && typeof this.image.scrollIntoView === 'function')) {
      this.image.scrollIntoView();  // Need to find a way to test this...
    }
  }

  onSelectImage() {
    this.props.onSelectImage(this.props.index);
  }

  render() {
    return (
      <div id={this.props.id} ref={(c) => { this.image = c; }} className={classnames({ image: true, active: this.props.isActive })} onClick={this.onSelectImage}>
        <img alt="Thumbnail" src={this.props.image.thumbnail} />
      </div>
    );
  }
}
