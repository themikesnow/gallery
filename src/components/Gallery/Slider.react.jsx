import React, { Component } from 'react';
import classnames from 'classnames';
import Loading from '../Loading/Loading.react';
import ImageProps from '../../constants/PropTypes';

export default class Slider extends Component {

  static propTypes = {
    image: React.PropTypes.shape(ImageProps),
    isPreviousEnabled: React.PropTypes.bool,
    isNextEnabled: React.PropTypes.bool,
    onPreviousImage: React.PropTypes.func.isRequired,
    onNextImage: React.PropTypes.func.isRequired,
  };

  static defaultProps = {
    image: null,
    isPreviousEnabled: false,
    isNextEnabled: false,
  };

  constructor(props) {
    super(props);
    this.onLoad = ::this.onLoad;
  }

  state = {
    isLoading: true,
  };

  componentWillReceiveProps(nextProps) {
    if (this.props.image !== nextProps.image) {
      this.setState({ isLoading: true });
    }
  }

  onLoad() {
    setTimeout(() => { // Added delay to make sure loading component is displayed...
      this.setState({ isLoading: false });
    }, 500);
  }

  render() {
    if (!this.props.image) {
      return <span />;
    }

    return (
      <div className="slider">
        <button className={classnames({ 'btn btn-previous': true, disabled: !this.props.isPreviousEnabled })} onClick={this.props.onPreviousImage}>
          Previuos
        </button>
        <button className={classnames({ 'btn btn-next': true, disabled: !this.props.isNextEnabled })} onClick={this.props.onNextImage}>
          Next
        </button>

        {this.state.isLoading && this.props.image.src && <Loading />}
        {this.props.image.src ?
          <img
            alt="Slider"
            src={this.props.image.src}
            onLoad={this.onLoad}
            className={classnames({ hidden: this.state.isLoading })}
          />
          :
          <div className="image-not-found">
            <i className="fa fa-exclamation-triangle" aria-hidden="true" />
            <p className="text">Unable to load image...</p>
          </div>
        }
      </div>
    );
  }
}
