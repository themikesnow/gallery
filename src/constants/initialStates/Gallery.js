
export default {
  searchText: '',
  selectedImage: 0,
  isBusy: false,
  images: null,
  isPreviousEnabled: false,
  isNextEnabled: false,
  responseDetails: {},
};
