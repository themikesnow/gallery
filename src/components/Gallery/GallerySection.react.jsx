import React, { Component } from 'react';
import { connect } from 'react-redux';
import { setSelectedImage, search, goToPreviousImage, goToNextImage, fetchMoreImagesIfNeeded } from '../../actions';
import Gallery from './Gallery.react';
import ImageProps from '../../constants/PropTypes';


export class GallerySection extends Component {

  static propTypes = {
    images: React.PropTypes.arrayOf(React.PropTypes.shape(ImageProps)),
    selectedImage: React.PropTypes.number,
    isPreviousEnabled: React.PropTypes.bool,
    isNextEnabled: React.PropTypes.bool,
    isBusy: React.PropTypes.bool,
    dispatch: React.PropTypes.func.isRequired,
  };

  static defaultProps = {
    isBusy: false,
    images: null,
    selectedImage: 0,
    isPreviousEnabled: false,
    isNextEnabled: false,
  };

  constructor(props) {
    super(props);
    this.onSelectImage = ::this.onSelectImage;
    this.onSearch = ::this.onSearch;
    this.onPreviousImage = ::this.onPreviousImage;
    this.onNextImage = ::this.onNextImage;
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.selectedImage !== this.props.selectedImage) {
      const { dispatch } = nextProps;
      dispatch(fetchMoreImagesIfNeeded());
    }
  }

  onSelectImage(image) {
    const { dispatch } = this.props;
    dispatch(setSelectedImage(image));
  }

  onSearch(text) {
    this.props.dispatch(search(text));
  }

  onPreviousImage() {
    this.props.dispatch(goToPreviousImage());
  }

  onNextImage() {
    this.props.dispatch(goToNextImage());
  }

  render() {
    return (
      <div className="gallery-section">
        <Gallery
          images={this.props.images}
          selectedImage={this.props.selectedImage}
          isPreviousEnabled={this.props.isPreviousEnabled}
          isNextEnabled={this.props.isNextEnabled}
          isBusy={this.props.isBusy}
          onSelectImage={this.onSelectImage}
          onSearch={this.onSearch}
          onNextImage={this.onNextImage}
          onPreviousImage={this.onPreviousImage}
        />
      </div>
    );
  }
}


const mapStateToProps = (state) => {
  const { isBusy, images, selectedImage, isPreviousEnabled, isNextEnabled, searchText } = state;

  return {
    isBusy,
    images,
    selectedImage,
    isPreviousEnabled,
    isNextEnabled,
    searchText,
  };
};

export default connect(mapStateToProps)(GallerySection);
