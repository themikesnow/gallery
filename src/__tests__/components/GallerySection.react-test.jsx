import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { mount } from 'enzyme';
import { expect as expectChai } from 'chai';
import sinon from 'sinon';

import InitialState from '../../constants/initialStates/Gallery';
import { Images } from '../data/Gallery';


import GallerySectionContainer, { GallerySection } from '../../components/Gallery/GallerySection.react';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
InitialState.images = Images;
const store = mockStore(InitialState);

describe('<GallerySection />', () => {
  it('Match Snapshot', () => {
    const component = renderer.create(
      <Provider store={store}>
        <GallerySectionContainer />
      </Provider>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should call onSearch when enter pressed on the search input', () => {
    sinon.spy(GallerySection.prototype, 'onSearch');
    const wrapper = mount(
      <Provider store={store}>
        <GallerySectionContainer />
      </Provider>,
    );
    wrapper.find('.btn-submit').simulate('click');
    expectChai(GallerySection.prototype.onSearch).to.have.property('callCount', 1);
  });

  it('should call onSelectImage when image is clicked', () => {
    sinon.spy(GallerySection.prototype, 'onSelectImage');
    const wrapper = mount(
      <Provider store={store}>
        <GallerySectionContainer />
      </Provider>,
    );
    wrapper.find('#image-0').simulate('click');
    expectChai(GallerySection.prototype.onSelectImage).to.have.property('callCount', 1);
  });

  it('should call onPreviousImage when previuos button is clicked', () => {
    sinon.spy(GallerySection.prototype, 'onPreviousImage');
    const wrapper = mount(
      <Provider store={store}>
        <GallerySectionContainer />
      </Provider>,
    );
    wrapper.find('.btn-previous').simulate('click');
    expectChai(GallerySection.prototype.onPreviousImage).to.have.property('callCount', 1);
  });

  it('should call onNextImage when next button is clicked', () => {
    sinon.spy(GallerySection.prototype, 'onNextImage');
    const wrapper = mount(
      <Provider store={store}>
        <GallerySectionContainer />
      </Provider>,
    );
    wrapper.find('.btn-next').simulate('click');
    expectChai(GallerySection.prototype.onNextImage).to.have.property('callCount', 1);
  });
});
