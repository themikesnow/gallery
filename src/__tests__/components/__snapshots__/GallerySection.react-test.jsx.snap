// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GallerySection /> Match Snapshot 1`] = `
<div
  className="gallery-section"
>
  <div
    className="gallery"
  >
    <div
      className="col-centered REVIEW"
    >
      <div
        className="search-input"
      >
        <div
          className="controls"
        >
          <input
            className="search-input-control"
            onChange={[Function]}
            onKeyPress={[Function]}
            placeholder="Enter keyword..."
            type="text"
            value=""
          />
          <button
            className="btn btn-submit disabled"
            onClick={[Function]}
          >
            <span>
              Search
            </span>
          </button>
        </div>
      </div>
    </div>
    <div>
      <div>
        <div
          className="slider"
        >
          <button
            className="btn btn-previous disabled"
            onClick={[Function]}
          >
            Previuos
          </button>
          <button
            className="btn btn-next disabled"
            onClick={[Function]}
          >
            Next
          </button>
          <div
            className="loading"
          >
            <i
              className="fa fa-refresh fa-spin fa-3x fa-fw"
            />
          </div>
          <img
            alt="Slider"
            className="hidden"
            onLoad={[Function]}
            src="https://dummy.com/123/123_o.jpg"
          />
        </div>
        <div
          className="images"
        >
          <div
            className="image active"
            id="image-0"
            onClick={[Function]}
          >
            <img
              alt="Thumbnail"
              src="https://dummy.com/123/123_q.jpg"
            />
          </div>
          <div
            className="image"
            id="image-1"
            onClick={[Function]}
          >
            <img
              alt="Thumbnail"
              src="https://dummy.com/456/456_q.jpg"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
